ff/serving/runner/materialize.go

```diff
@@ -0,0 +1,148 @@
+package runner
+
+import (
+	"fmt"
+	provider "github.com/featureform/serving/provider"
+)
+
+const MAXIMUM_CHUNK_ROWS int64 = 1024
+const WORKER_IMAGE string = "featureform/worker"
+
+type JobCloud string
+
+const (
+	Kubernetes JobCloud = "KUBERNETES"
+	Local      JobCloud = "LOCAL"
+)
+
+type MaterializeRunner struct {
+	Online  provider.OnlineStore
+	Offline provider.OfflineStore
+	ID      provider.ResourceID
+	Cloud   JobCloud
+}
+
+type WatcherMultiplex struct {
+	CompletionList []CompletionWatcher
+}
+
+func (w WatcherMultiplex) Complete() bool {
+	complete := true
+	for _, completion := range w.CompletionList {
+		complete = complete && completion.Complete()
+	}
+	return complete
+}
+func (w WatcherMultiplex) String() string {
+	complete := 0
+	for _, completion := range w.CompletionList {
+		if completion.Complete() {
+			complete += 1
+		}
+	}
+	return fmt.Sprintf("%v complete out of %v", complete, len(w.CompletionList))
+}
+func (w WatcherMultiplex) Wait() error {
+	for _, completion := range w.CompletionList {
+		if err := completion.Wait(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func (w WatcherMultiplex) Err() error {
+	for _, completion := range w.CompletionList {
+		if err := completion.Err(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m MaterializeRunner) Run() (CompletionWatcher, error) {
+	materialization, err := m.Offline.CreateMaterialization(m.ID)
+	if err != nil {
+		return nil, err
+	}
+	chunkSize := MAXIMUM_CHUNK_ROWS
+	var numChunks int64
+	numRows, err := materialization.NumRows()
+	if err != nil {
+		return nil, err
+	}
+	if numRows <= MAXIMUM_CHUNK_ROWS {
+		chunkSize = numRows
+		numChunks = 1
+	} else if chunkSize == 0 {
+		numChunks = 0
+	} else if numRows > chunkSize {
+		numChunks = numRows / chunkSize
+		if chunkSize*numChunks < numRows {
+			numChunks += 1
+		}
+	}
+	config := &MaterializedChunkRunnerConfig{
+		OnlineType:     m.Online.Type(),
+		OfflineType:    m.Offline.Type(),
+		OnlineConfig:   m.Online.Config(),
+		OfflineConfig:  m.Offline.Config(),
+		MaterializedID: materialization.ID(),
+		ResourceID:     m.ID,
+		ChunkSize:      chunkSize,
+	}
+	serializedConfig, err := config.Serialize()
+	if err != nil {
+		return nil, err
+	}
+	var cloudWatcher CompletionWatcher
+	switch m.Cloud {
+	case Kubernetes:
+		envVars := map[string]string{"NAME": "COPY", "CONFIG": string(serializedConfig)}
+		kubernetesConfig := KubernetesRunnerConfig{
+			envVars:  envVars,
+			image:    WORKER_IMAGE,
+			numTasks: int32(numChunks),
+		}
+		kubernetesRunner, err := NewKubernetesRunner(kubernetesConfig)
+		if err != nil {
+			return nil, err
+		}
+		cloudWatcher, err = kubernetesRunner.Run()
+		if err != nil {
+			return nil, err
+		}
+	case Local:
```

Comment:
I would have made LocalMaterializeRunner and KubernetesMaterializeRunner, this is definitely a case for different classes.

Reply:
changed names